include/ginkgo/core/config/registry.hpp

```diff
@@ -0,0 +1,220 @@
+     */
+    explicit type_descriptor(std::string value_typestr = "void",
+                             std::string index_typestr = "void");
+
+    /**
+     * Get the value type string.
+     */
+    const std::string& get_value_typestr() const;
+
+    /**
+     * Get the index type string
+     */
+    const std::string& get_index_typestr() const;
+
+private:
+    std::string value_typestr_;
+    std::string index_typestr_;
+};
+
+
+template <typename ValueType = void, typename IndexType = void>
+type_descriptor make_type_descriptor();
+
+
+using linop_map = std::unordered_map<std::string, std::shared_ptr<LinOp>>;
+using linopfactory_map =
+    std::unordered_map<std::string, std::shared_ptr<LinOpFactory>>;
+using criterionfactory_map =
+    std::unordered_map<std::string, std::shared_ptr<stop::CriterionFactory>>;
+using buildfromconfig_map =
+    std::map<std::string,
+             std::function<deferred_factory_parameter<gko::LinOpFactory>(
+                 const pnode&, const registry&, type_descriptor)>>;
+
+
+/**
+ * registry is the storage for file config usage. It stores the building
+ * function, linop, linop_factory, criterion
+ */
+class registry {
+private:
+    /**
+     * map_type gives the map type according to the base type of given type.
+     *
+     * @tparam T  the type
+     */
+    template <typename T, typename = void>
+    struct map_type {
+        using type = void;
+    };
+
+    template <typename T>
+    struct map_type<T, typename std::enable_if<
+                           std::is_convertible<T*, LinOp*>::value>::type> {
+        using type = linop_map;
+    };
+
+    template <typename T>
+    struct map_type<T, typename std::enable_if<std::is_convertible<
+                           T*, LinOpFactory*>::value>::type> {
+        using type = linopfactory_map;
+    };
+
+    template <typename T>
+    struct map_type<T, typename std::enable_if<std::is_convertible<
+                           T*, stop::CriterionFactory*>::value>::type> {
+        using type = criterionfactory_map;
+    };
+
+public:
+    /**
+     * registry constructor
+     *
+     * @param build_map  the build map to dispatch the class base. Ginkgo
+     * provides `generate_config_map()` in config.hpp to provide the ginkgo
+     * build map. Users can extend this map to fit their own LinOpFactory.
+     */
+    registry(buildfromconfig_map build_map) : build_map_(build_map) {}
+
+    /**
+     * insert_data stores the data with the key.
+     *
+     * @tparam T  the type
+     *
+     * @param key  the unique key string
+     * @param data  the shared pointer of the object
+     */
+    template <typename T>
+    bool emplace(std::string key, std::shared_ptr<T> data);
+
+    /**
+     * search_data searches the key on the corresponding map.
+     *
+     * @tparam T  the type
+     *
+     * @param key  the key string
+     *
+     * @return the shared pointer of the object
+     */
+    template <typename T>
```

Comment:
I think this doesn't need to be a template.
At the time of configuring, we should always know if we expect a linop, a linopfactory, or a criterionfactory.

Reply:
It's more than that. In some cases, we asked more specific type than linop/linopfactory.

Reply:
Right, then it's fine to keep it.